uplink/ecclient/client.go

```diff
@@ -473,30 +474,81 @@ func (lr *lazyPieceRanger) Size() int64 {
 // Range implements Ranger.Range to be lazily connected
 func (lr *lazyPieceRanger) Range(ctx context.Context, offset, length int64) (_ io.ReadCloser, err error) {
 	defer mon.Task()(&ctx)(&err)
-	ps, err := lr.dialPiecestore(ctx, &pb.Node{
+
+	reader := &lazyPieceReader{}
+	reader.lazyPieceRanger = lr
+	reader.ctx = ctx
+	reader.offset = offset
+	reader.length = length
+	return reader, nil
+}
+
+type lazyPieceReader struct {
+	*lazyPieceRanger
+	ctx    context.Context
+	offset int64
+	length int64
+
+	mu sync.RWMutex
+
+	dialed bool
+	closed bool
+	piecestore.Downloader
+	client *piecestore.Client
+}
+
+func (lr *lazyPieceReader) Read(data []byte) (_ int, err error) {
+	lr.mu.RLock()
+	defer lr.mu.RUnlock()
+
+	if lr.closed {
+		return 0, io.EOF
+	}
+	if !lr.dialed {
+		err = lr.dial()
+		if err != nil {
+			return 0, err
+		}
+	}
+	if lr.dialed && lr.Downloader == nil {
+		return 0, io.EOF
+	}
+
+	return lr.Downloader.Read(data)
+}
+
+func (lr *lazyPieceReader) dial() (err error) {
+	defer mon.Task()(&lr.ctx)(&err)
+	lr.dialed = true
+	ps, err := lr.dialPiecestore(lr.ctx, &pb.Node{
 		Id:      lr.limit.GetLimit().StorageNodeId,
 		Address: lr.limit.GetStorageNodeAddress(),
 	})
 	if err != nil {
-		return nil, err
+		return err
 	}
+	lr.client = ps
```

Comment:
It looks like the only reason we're holding on to lr.client is to close it on a close call. I suggest we add a field on `lr` called `closers`, like:

`closers []io.Closer`

Then, you add each:

```
lr.closers = append(lr.closers, downloader)
lr.cloers = append(lr.closers, client)
```

Then in the close function:

```
for _, closer := range lr.closers {
    err = errs.Combine(err, closer.Close())
}
return err
```

You know, now that I've written it out I feel more ambivalent about it. I'm going to leave the comment so you can see the idea, the nice thing is you're not holding onto specific member variables that only exist to be closed (at the risk of forgetting later!). But I'm sixes on it.